cilium-health/launch/endpoint.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package launch
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/common/addressing"
+	"github.com/cilium/cilium/common/plugins"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/vishvananda/netlink"
+)
+
+func logFromCommand(cmd *exec.Cmd, netns string) error {
+	scopedLog := log.WithField("netns", netns)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stdout)
+		for in.Scan() {
```

Comment:
It's usually better to keep the output of one command in a single log statement. Otherwise, the output may interleave and make the logs very hard to read. One way to do is to concatenate the lines with newlines or some special character (`"\\n"`) in the same log message.

Reply:
This is reading stdout from a long-running command, using the default `ScanLines` implementation, so each time the cilium-health endpoint logs a message, it will be forwarded to the Cilium log. I'm not sure how we can achieve what you're saying.

Reply:
@joestringer is right about the default [`ScanLines`](https://golang.org/pkg/bufio/#ScanLines) implementation. But reading the documentation I understand @rlenglet concerns. The problem might be with "the last non-empty line of input will be returned even if it has no newline" in which case it will log something without a `\n`. I don't think this is a real issues since logrus will print a log message per method call (`Debugf`, `Infof`, etc) even if it ends with or without `\n`

Reply:
OK, makes sense. LGTM.